### DISCUSSION TITLE
Handling desynchronized map/preset

### DISCUSSION THREAD

**Discussion by PiotrMachowski** [PiotrMachowski/lovelace-xiaomi-vacuum-map-card#319] [FAQ]
If you use `activate` with `activate_on_switch` to switch presented maps and current map preset becomes desynchronized from shown map you can just click on preset's name to manually trigger `activate` service call.

![image](https://user-images.githubusercontent.com/6118709/149677630-8b5a743b-edc7-4c58-b20a-4e3b0736775e.png)

**Comment by eSKape-debug** [comment]
@PiotrMachowski is it somehow possible to call this `activate` service from a homeassistant script or automation?
I'd like to load the current preset (matching the map) of my 3 presets the time the card is shown - while not loosing the functionality of preset switching as shown in the picture above.

**Reply by PiotrMachowski** [reply]
`activate` is just a name of a section in the config. It contains a normal service call, so you can normally use it in scripts/automations

**Reply by junado** [reply]
I think what eSKape would like to do is have an HA script/automation select a preset.

In my case, I would use it to ask the card to select the proper map upon refreshing my HA dashboard. I use presets for a multi-level home and if I refresh, the card always reverts to the first preset, which does not always match the floor the robot is on. I can figure out which floor the robot is on, but I can't see a way to have the vacuum card select the preset to match.

**Reply by PiotrMachowski** [reply]
It is possible to add conditions to presets - map card will display appropriate presets accordingly

**Reply by junado** [reply]
From my testing, this hides or show the presets, but can't let me choose one if multiple presets are shown. It could work, but would prevent me from switching presets if I only show the active one, unless I am missing something. I could do this if I use a script to switch floors, but would disable the vacuum card's innate ability to switch presets.

**Reply by PiotrMachowski** [reply]
It is possible to add icons that will change maps on the vacuum and keep presets automatic with conditions.

**Reply by junado** [reply]
That would work, but it feels like a workaround.

I appreciate the work you put into the vacuum card by the way, please don't see this as a complaint !

**Reply by PiotrMachowski** [reply]
What approach would you consider as a correct one? Just selecting the preset based on the condition, instead of hiding it?

**Reply by eSKape-debug** [reply]
i fixed that for myself using the changes here: https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/compare/master...eSKape-debug:lovelace-xiaomi-vacuum-map-card:master

a new top level config param is introduced which uses an input_number entity to store the currently selected map id - however, as this was just quick and dirty, this currently only works if the order of presets is the same as the numerical order of map ids, meaning 1st preset is id 0, 2nd preset is id 1, etc. (array index selection in code)

```javascript
type: custom:xiaomi-vacuum-map-card
preset_name: Erdgeschoss
entity: vacuum.roborock_vacuum_s5e
current_preset: input_number.vacuum_map_id
activate_on_switch: true
activate:
  service: script.turn_on
  service_data:
    entity_id: script.vacuum_handle_map_reload
    variables:
      map_id: 0
.....
``` 

the script in the actions additionally handles the update of the entity (might also be done in the map-card code)
As a first step - this works for me

Screencast of the result: https://github.com/eSKape-debug/lovelace-xiaomi-vacuum-map-card/blob/master/docs/Bildschirmaufnahme%202024-02-22%20um%2019.32.08-1.mov

**Reply by junado** [reply]
If it has to work within the confines of the yaml card configuration, perhaps having a "selected" true/false option as part of the preset options? I'd see it as optional, defaults to false. If multiple values are true, select the first preset that matches. If all values are false, defaults to first preset (current operation). This would let us add conditional selection of active preset through Jinja templates.

**Reply by PiotrMachowski** [reply]
@junado keep in mind that jinja templates are not currently allowed in yaml config - that's why I used conditions instead.

**Reply by junado** [reply]
> @junado keep in mind that jinja templates are not currently allowed in yaml config - that's why I used conditions instead.

I understand, I meant JavaScript (it's a bit confusing, since Jinja can be used inside of styles for card_mod, so it's all mixed in there). Since you already have the conditions built into the card, perhaps using conditions to select the active preset could work? It would essentially be the same as the show/hide conditions, but would select the preset if the condition is true.
